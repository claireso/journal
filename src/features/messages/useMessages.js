import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

const MessagesContext = React.createContext()

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

export const MessagesProvider = ({ children }) => {
  const [state, setState] = useState([])

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = () => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status) => (message) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
