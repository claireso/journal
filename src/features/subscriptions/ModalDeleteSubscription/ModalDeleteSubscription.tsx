import { useCallback } from 'react'

import { ButtonPrimary, ButtonSecondary } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Text from '@components/Text'

interface ModalDeleteSubscriptionProps {
  id: number
  onConfirm: (id: number) => void
  onCancel: () => void
  isProcessing?: boolean
}

const ModalDeleteSubscription = ({ onConfirm, onCancel, isProcessing = false, id }: ModalDeleteSubscriptionProps) => {
  const onClickConfirm = useCallback(() => {
    onConfirm(id)
  }, [id, onConfirm])

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <ButtonSecondary onClick={onCancel}> Cancel </ButtonSecondary>
        <ButtonPrimary onClick={onClickConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </ButtonPrimary>
      </Text>
    </>
  )
}

export default ModalDeleteSubscription
