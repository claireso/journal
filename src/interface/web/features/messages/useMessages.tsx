'use client'

import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

interface Message {
  key?: string
  status: 'default' | 'success' | 'error' | 'info'
  message: string
}

interface Actions {
  displaySuccessMessage: (data: Omit<Message, 'status'>) => void
  displayErrorMessage: (data: Omit<Message, 'status'>) => void
  closeMessage: (index: number) => void
  closeAllMessages: () => void
}

const INITIAL_STATE: Message[] = []

type TMessagesContext<T> = [T, React.Dispatch<React.SetStateAction<T>>]

const MessagesContext = React.createContext<TMessagesContext<Message[]>>([null!, null!])

interface MessagesProviderProps {
  children: React.ReactNode
}

export const MessagesProvider = ({ children }: MessagesProviderProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = (): [Message[], Actions] => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index: number) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status: Message['status']) => (message: Omit<Message, 'status'>) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message: Message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
