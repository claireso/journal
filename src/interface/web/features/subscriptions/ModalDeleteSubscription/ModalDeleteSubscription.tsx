import { useCallback } from 'react'

import { ButtonPrimary, ButtonSecondary } from '@web/components/Buttons'
import { Heading1 } from '@web/components/Headings'
import Text from '@web/components/Text'

interface ModalDeleteSubscriptionProps {
  id: number
  onConfirm: (id: number) => void
  onCancel: () => void
  isProcessing?: boolean
}

const ModalDeleteSubscription = ({ onConfirm, onCancel, isProcessing = false, id }: ModalDeleteSubscriptionProps) => {
  const onClickConfirm = useCallback(() => {
    onConfirm(id)
  }, [id, onConfirm])

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <ButtonSecondary onClick={onCancel}> Cancel </ButtonSecondary>
        <ButtonPrimary onClick={onClickConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </ButtonPrimary>
      </Text>
    </>
  )
}

export default ModalDeleteSubscription
