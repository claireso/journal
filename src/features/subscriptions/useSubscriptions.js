import React, { useState, useContext, useCallback } from 'react'
import PropTypes from 'prop-types'

import * as api from '@services/api'
import useMessages from '@features/messages/useMessages'

let currentRequest = null

const SubscriptionsContext = React.createContext()

const initialState = {
  subscriptions: {},
  isLoading: true,
  isProcessing: false
}

const useSubscriptionsContext = () => {
  const context = useContext(SubscriptionsContext)

  if (context === undefined) {
    throw new Error('useSubscriptionsContext must be used within a SubscriptionsProvider')
  }

  return context
}

export const SubscriptionsProvider = ({ value = initialState, children }) => {
  const [state, setState] = useState(value)

  return <SubscriptionsContext.Provider value={[state, setState]}>{children}</SubscriptionsContext.Provider>
}

SubscriptionsProvider.propTypes = {
  children: PropTypes.any,
  value: PropTypes.object
}

const useSubscriptions = () => {
  const [, { displaySuccessMessage, displayErrorMessage }] = useMessages()
  const [state, setState] = useSubscriptionsContext()
  const { subscriptions, isLoading, isProcessing } = state

  const updateState = (newState) => setState({ ...state, ...newState })

  const loadSubscriptions = useCallback(async (page = 1) => {
    if (currentRequest) {
      currentRequest.abort()
    }

    currentRequest = api.getSubscriptions(page)

    try {
      const response = await currentRequest.ready
      currentRequest = null
      updateState({
        subscriptions: {
          ...response
        },
        isLoading: false
      })
    } catch {
      updateState({ isLoading: false })
    }
  }, [])

  const deleteSubscription = useCallback(async (id) => {
    try {
      updateState({ isProcessing: true })
      await api.deleteSubscription(id).ready
      updateState({
        isProcessing: false,
        subscriptions: {
          items: subscriptions.items.filter((subscription) => subscription.id !== id),
          pager: {
            ...subscriptions.pager,
            count: subscriptions.pager.count - 1
          }
        }
      })
      displaySuccessMessage({
        message: 'Your subscription has been deleted successfully',
        key: 'CRUD_SUBSCRIPTION'
      })
    } catch {
      updateState({ isProcessing: false })
      displayErrorMessage({
        message: 'An error has occured during the deletion. Please retry',
        key: 'CRUD_SUBSCRIPTION'
      })
    }
  }, [])

  return [
    {
      data: subscriptions.items,
      pager: subscriptions.pager,
      isLoading,
      isProcessing
    },
    {
      loadSubscriptions,
      deleteSubscription
    }
  ]
}

export default useSubscriptions
