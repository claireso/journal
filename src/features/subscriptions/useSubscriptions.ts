import { useQuery, useQueryClient, useMutation, QueryFunctionContext } from '@tanstack/react-query'

import { Subscriptions, Subscription } from '@models'
import * as api from '@services/api'
import useMessages from '@features/messages/useMessages'

interface Filters {
  page: string
}

const initialState: Subscriptions = {
  items: [],
  pager: {
    count: 0
  }
}

const CACHE_KEY_LIST = 'subscriptions'

/**
 * Delete subscription and update cache
 * @returns object
 */
export const useDeleteSubscription = (filters: Filters) => {
  const queryClient = useQueryClient()
  const [, { displaySuccessMessage, displayErrorMessage }] = useMessages()

  return useMutation({
    mutationFn: (id: Subscription['id']) => api.deleteSubscription(id),
    onSuccess(data, id) {
      const subscriptions = queryClient.getQueryData<Subscriptions>([CACHE_KEY_LIST, filters])

      if (subscriptions) {
        queryClient.setQueryData([CACHE_KEY_LIST, filters], {
          items: subscriptions.items.filter((subscription) => subscription.id !== id),
          pager: {
            ...subscriptions.pager,
            count: subscriptions.pager.count - 1
          }
        })
      }

      displaySuccessMessage({
        key: 'CRUD_SUBSCRIPTION',
        message: 'Your subscription has been deleted successfully'
      })
    },
    onError() {
      displayErrorMessage({
        key: 'CRUD_SUBSCRIPTION',
        message: 'An error has occured during the deletion. Please retry'
      })
    }
  })
}

/**
 * Fetch subscriptions
 */
export const useSubscriptions = (filters: Filters = { page: '1' }, options = {}) => {
  const getSubscriptions = async ({ signal }: QueryFunctionContext) =>
    await api.getSubscriptions(filters.page, { signal })

  const queryOptions = {
    placeholderData: initialState,
    ...options
  }

  return useQuery({ queryKey: [CACHE_KEY_LIST, filters], queryFn: getSubscriptions, ...queryOptions })
}
