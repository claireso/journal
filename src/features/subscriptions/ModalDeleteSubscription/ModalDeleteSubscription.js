import PropTypes from 'prop-types'

import { PrimaryButton, SecondaryButton } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Text from '@components/Text'

import useSubscriptions from '@features/subscriptions/useSubscriptions'

const ModalDeleteSubscription = ({ onClose, id }) => {
  const [{ isProcessing }, { deleteSubscription }] = useSubscriptions()

  const onCancel = () => onClose()

  const onConfirm = async () => {
    await deleteSubscription(id)
    onClose()
  }

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <SecondaryButton onClick={onCancel}> Cancel </SecondaryButton>
        <PrimaryButton onClick={onConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </PrimaryButton>
      </Text>
    </>
  )
}

ModalDeleteSubscription.propTypes = {
  id: PropTypes.number.isRequired,
  onClose: PropTypes.func
}

ModalDeleteSubscription.defaultProps = {
  onClose: () => {}
}

export default ModalDeleteSubscription
