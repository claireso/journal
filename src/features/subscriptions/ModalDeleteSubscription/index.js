import ModalDeleteSubscription from './ModalDeleteSubscription'

export default ModalDeleteSubscription
