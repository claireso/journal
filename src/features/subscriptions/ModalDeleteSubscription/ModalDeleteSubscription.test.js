import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { SubscriptionsProvider } from '../useSubscriptions'
import { MessagesProvider } from '../../messages/useMessages'
import ModalDeleteSubscription from './ModalDeleteSubscription'

import * as api from '@services/api'

describe('<ModalDeleteSubscription />', () => {
  const renderComponent = (props) =>
    render(
      <MessagesProvider>
        <SubscriptionsProvider>
          <ModalDeleteSubscription id={188} {...props} />
        </SubscriptionsProvider>
      </MessagesProvider>
    )

  beforeEach(() => {
    jest.spyOn(api, 'deleteSubscription')
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render component', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not delete subscription', () => {
    const props = {
      onClose: jest.fn()
    }

    renderComponent(props)

    fireEvent.click(screen.getByText('Cancel'))

    expect(props.onClose).toHaveBeenCalled()
    expect(api.deleteSubscription).not.toHaveBeenCalled()
  })

  it('should delete subscription', async () => {
    const props = {
      onClose: jest.fn()
    }

    renderComponent(props)

    fireEvent.click(screen.getByText('Yes'))

    expect(api.deleteSubscription).toHaveBeenCalledWith(188)

    await waitFor(() => expect(props.onClose).toHaveBeenCalled())
  })
})
