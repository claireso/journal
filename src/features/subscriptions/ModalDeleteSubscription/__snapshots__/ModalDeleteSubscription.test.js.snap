// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDeleteSubscription /> should render component 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 2.4rem;
  margin: 0 0 2.4rem;
  font-weight: 500;
}

<h1
    class="c0"
  >
    Are you sure?
  </h1>
  <p>
    This action is irreversible
  </p>
  .c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
}

.c3 {
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c3 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c1 + .c1 {
  margin: 0 0 0 0.8rem;
}

.c6 {
  background: var(--primary-normal);
  color: white;
}

.c6:hover {
  background: var(--primary-lighter);
}

.c4 {
  background: var(--secondary-normal);
  color: var(--text-normal);
}

.c4:hover {
  background: var(--secondary-darker);
}

.c4 + .c5,
.c5 + .c4,
.c4 + .c4 {
  margin-left: 0.8rem;
}

.c0 {
  text-align: right;
  font-size: var(--font-size-smaller);
}

<p
    class="c0"
  >
    <button
      class="c1 c2 c3 c4"
    >
       Cancel 
    </button>
    <button
      class="c1 c5 c2 c3 c6"
    >
       Yes 
    </button>
  </p>
</DocumentFragment>
`;
